src/transformers/trainer.py

```diff
@@ -782,56 +785,56 @@ def create_optimizer(self):
-                {
-                    "params": [p for n, p in self.model.named_parameters() if n in decay_parameters],
-                    "weight_decay": self.args.weight_decay,
-                },
-                {
-                    "params": [p for n, p in self.model.named_parameters() if n not in decay_parameters],
-                    "weight_decay": 0.0,
-                },
-            ]
-            optimizer_cls = Adafactor if self.args.adafactor else AdamW
-            if self.args.adafactor:
-                optimizer_cls = Adafactor
-                optimizer_kwargs = {"scale_parameter": False, "relative_step": False}
-            else:
-                optimizer_cls = AdamW
-                optimizer_kwargs = {
-                    "betas": (self.args.adam_beta1, self.args.adam_beta2),
-                    "eps": self.args.adam_epsilon,
-                }
-            optimizer_kwargs["lr"] = self.args.learning_rate
-            if self.sharded_ddp == ShardedDDPOption.SIMPLE:
-                self.optimizer = OSS(
-                    params=optimizer_grouped_parameters,
-                    optim=optimizer_cls,
-                    **optimizer_kwargs,
-                )
-            else:
-                self.optimizer = optimizer_cls(optimizer_grouped_parameters, **optimizer_kwargs)
+        decay_parameters = get_parameter_names(self.model, [nn.LayerNorm])
+        decay_parameters = [name for name in decay_parameters if "bias" not in name]
+        optimizer_grouped_parameters = [
+            {
+                "params": [p for n, p in self.model.named_parameters() if n in decay_parameters],
+                "weight_decay": self.args.weight_decay,
+            },
+            {
+                "params": [p for n, p in self.model.named_parameters() if n not in decay_parameters],
+                "weight_decay": 0.0,
+            },
+        ]
+        optimizer_cls = Adafactor if self.args.adafactor else AdamW
+        if self.args.adafactor:
+            optimizer_cls = Adafactor
+            optimizer_kwargs = {"scale_parameter": False, "relative_step": False}
+        else:
+            optimizer_cls = AdamW
+            optimizer_kwargs = {
+                "betas": (self.args.adam_beta1, self.args.adam_beta2),
+                "eps": self.args.adam_epsilon,
+            }
+        optimizer_kwargs["lr"] = self.args.learning_rate
+        if self.sharded_ddp == ShardedDDPOption.SIMPLE:
+            optimizer = OSS(
+                params=optimizer_grouped_parameters,
+                optim=optimizer_cls,
+                **optimizer_kwargs,
+            )
+        else:
+            optimizer = optimizer_cls(optimizer_grouped_parameters, **optimizer_kwargs)
 
         if is_sagemaker_mp_enabled():
-            self.optimizer = smp.DistributedOptimizer(self.optimizer)
+            optimizer = smp.DistributedOptimizer(optimizer)
 
-    def create_scheduler(self, num_training_steps: int):
+        return optimizer
```

Comment:
```suggestion
        self.optimizer = optimizer
        return optimizer
```

this breaks back-compatibility so it needs to be changed back to keep `create_optimizer` setting ` self.optimizer` internally and the same for `create_scheduler`. So the only change needed is to set the attributes beside returning to keep the functions work as before.

Reply:
Thanks for the feedback. The previous code also had guards like `if self.optimizer is None:` to prevent repeated execution of the optimizer and scheduler creation logic. So I will revert back to the old code and only add the return optimizer and scheduler statements.

Reply:
Ah yes! I guess it was used primarily internally initially so that's why it probably wasn't designed to be a user-level function.

Your suggestion to keep things as before and just add a return is probably the best. We will also run this by Sylvain once things are ready.